crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs

```diff
@@ -701,3 +707,233 @@ pub(crate) fn numpy_2_0_deprecation(checker: &mut Checker, expr: &Expr) {
+                return false;
+            }
+            if !binding.in_exception_handler() {
+                return false;
+            }
+            let Some(try_node) = binding
+                .source
+                .and_then(|import_id| enclosing_try_node(import_id, semantic))
+            else {
+                return false;
+            };
+            let suspended_exceptions = Exceptions::from_try_stmt(try_node, semantic);
+            if !suspended_exceptions
+                .intersects(Exceptions::IMPORT_ERROR | Exceptions::MODULE_NOT_FOUND_ERROR)
+            {
+                return false;
+            }
+            try_block_contains_undeprecated_import(try_node, &replacement.details)
+        }
+        _ => false,
+    }
+}
+
+/// Given the `NodeId` of a node that is known to be inside an exception handler,
+/// find the [`ast::StmtTry`] node in which the node resides.
+fn enclosing_try_node<'a>(
+    node_id: NodeId,
+    semantic: &'a SemanticModel,
+) -> Option<&'a ast::StmtTry> {
+    let mut node_id = node_id;
+    loop {
+        node_id = semantic.parent_statement_id(node_id)?;
+        if let ast::Stmt::Try(try_node) = semantic.statement(node_id) {
+            return Some(try_node);
+        }
+    }
+}
+
+/// Given an [`ast::StmtTry`] node, does the `try` branch of that node
+/// contain any [`ast::ExprAttribute`] nodes that indicate the numpy
+/// member is being accessed from the non-deprecated location?
+fn try_block_contains_undeprecated_attribute(
+    try_node: &ast::StmtTry,
+    replacement_details: &Details,
+    semantic: &SemanticModel,
+) -> bool {
+    let Details::AutoImport {
+        path,
+        name,
+        compatibility: _,
+    } = replacement_details
+    else {
+        return false;
+    };
+    let undeprecated_qualified_name = {
+        let mut builder = QualifiedNameBuilder::default();
+        for part in path.split('.') {
+            builder.push(part);
+        }
+        builder.push(name);
+        builder.build()
+    };
+    let mut attribute_searcher = AttributeSearcher::new(undeprecated_qualified_name, semantic);
+    for stmt in &try_node.body {
+        attribute_searcher.visit_stmt(stmt);
+        if attribute_searcher.found_attribute {
+            return true;
+        }
+    }
+    false
+}
+
+/// AST visitor that searches an AST tree for [`ast::ExprAttribute`] nodes
+/// that match a certain [`QualifiedName`].
+struct AttributeSearcher<'a> {
+    attribute_to_find: QualifiedName<'a>,
+    semantic: &'a SemanticModel<'a>,
+    found_attribute: bool,
+}
+
+impl<'a> AttributeSearcher<'a> {
+    fn new(attribute_to_find: QualifiedName<'a>, semantic: &'a SemanticModel<'a>) -> Self {
+        Self {
+            attribute_to_find,
+            semantic,
+            found_attribute: false,
+        }
+    }
+}
+
+impl Visitor<'_> for AttributeSearcher<'_> {
+    fn visit_expr(&mut self, expr: &'_ Expr) {
+        if self.found_attribute {
+            return;
+        }
```

Comment:
Nit: It could make sense to add a `visit_stmt` implementation that short-circuits visiting when `self.found_attribute` is true (or `visit_body`)

Reply:
Thanks, I pushed that change! Wasn't quite sure what best practices were here, since there's a lot of `visit_*` methods on the `Visitor` trait that could potentially be overridden here to short-circuit. I guess most of the other ones generally delegate to `visit_expr` and `visit_stmt`?